### PR TITLE
Add source code for Chapter 4

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Examples/NoModules/Chapter 04/Ex4_01A.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  std::cout << "Enter a character: ";
+  std::cin >> first;
+
+  std::cout << "Enter a second character: ";
+  std::cin >> second;
+
+  std::cout << std::format("The value of the expression {} < {} is {}\n",
+                           first, second, first < second);
+  std::cout << std::format("The value of the expression {} == {} is {}\n",
+                           first, second, first < second);
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 04/Ex4_01A.cpp:19]: oops!
first < second
moet zijn:
first == second
ook foutje in de text ... gemist tijdens review :(

File: Exercises/NoModules/Chapter 04/Soln4_02.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Exercise 4-1 Testing for exact division of one integer by another.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_02.cpp:1]: Exercise 4-2

File: Exercises/NoModules/Chapter 04/Soln4_02.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  if (value1 <= 0 || value2 <= 0)  // Valid input?
+  {
+    std::cout << "Sorry - positive integers only." << std::endl;
+    return 1;
+  }
+
+  // Ensure that value1 is not smaller than value2
+  if (value1 < value2)
+  {         
+    const auto temp = value1;    // swap if necessary
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_02.cpp:26]: uniform init

File: Exercises/NoModules/Chapter 04/Soln4_03.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
+// Exercise 4-2 Using nested ifs and a logical && to check the value of a number.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_03.cpp:1]: 4-3

File: Exercises/NoModules/Chapter 04/Soln4_04.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  char gender {};             // Initialized to '\0' (see Chapter 5)
+  AcademicDegree degree {};   // Initialized to AcademicDegree::none
+  bool married {};            // Initialized to false
+  bool speaksHindi {};
+  bool speaksUrdu {};
+
+  std::cout << "What is your age, if I may ask? ";
+  std::cin >> age;
+  
+  if (age < 0 || age > 130)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_04.cpp:24]: This solution doesn't seem to match the question.
The question states:
"People who enter invalid values do not qualify, of course, and should be ruled **out as early as possible (that is, immediately after entering any invalid value;**"

But in your solution, you first collect everything, then at the end you check everytrhing.

_[Edit: on second read, maybe it's ok and I miss interpreted the question. You are just validating input, not yet comparing to what we need.]_

Comment (MarcGregoire) [reply Exercises/NoModules/Chapter 04/Soln4_04.cpp:24]: Also, why not use unsigned int for age?

File: Exercises/NoModules/Chapter 04/Soln4_04.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  {
+    std::cout << "Sure it is, joker. Sadly, commedians don't qualify..." << std::endl;
+    return 1;
+  }
+  
+  std::cout << "What is your gender (m or f)? ";
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_04.cpp:30]: Only m and f?
These days there are lots of other options ...

File: Exercises/NoModules/Chapter 04/Soln4_04.cpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  std::cin >> gender;
+  
+  gender = std::tolower(gender);
+  if (gender != 'm' && gender != 'f')
+  {
+    std::cout << "We do not mean to discriminate, but our current software only accepts male or female. Sorry...";
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_04.cpp:36]: Maybe not such a good message these days ...

File: Exercises/NoModules/Chapter 04/Soln4_06.cpp [hunk 8]
```diff
@@ -0,0 +1,22 @@
+// Exercise 4-4 Finding the range for an integer.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_06.cpp:1]: 4-6

File: Exercises/NoModules/Chapter 04/Soln4_06.cpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
+// Exercise 4-4 Finding the range for an integer. 
+// This is just a question of bolting sufficent conditional operators together
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_06.cpp:2]: sufficient

File: Exercises/NoModules/Chapter 04/Soln4_07.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
+// Exercise 4-3 Outputting the binary code for a letter.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_07.cpp:1]: 4-7

File: Exercises/NoModules/Chapter 04/Soln4_07.cpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+/*
+ * Most of the program is fairly simple. 
+ * The cctype functions make determining upper or lower case easy. 
+ * Finding out if it's a vowel is also easy with a switch.
+ * Only getting the binary code needs a little thought, though.                                                           *
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_07.cpp:6]: what's with the * on the far right?

File: Exercises/NoModules/Chapter 04/Soln4_07.cpp [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      std::cout << "vowel";
+      break;
+    default:
+      std::cout << "consonant";
+      break;
+  }
+  std::cout << '.' << std::endl;
+
+  // Output the character code as binary
+  std::cout << "The binary code for '" << ch << "' is "
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_07.cpp:45]: The question says "output the lowercase letter ", you are not lowercasing ch.
If you do, maybe you can do ones, after inputting the letter, then you can remove the tolower in the switch above as well.

File: Exercises/NoModules/Chapter 04/Soln4_08.cpp [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  const unsigned dime {10};
+  const unsigned nickel {5};
+
+  double amountInDollars {0.0};
+  std::cout << std::endl << "Please enter a cash amount between 0 and 10 dollars: $";
+  std::cin >> amountInDollars;
+  
+  if (amountInDollars >= 0.0 && amountInDollars <= 10.0)
+  {
+    // Multiply dollar amount by 100 ($1 {100 cents)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_08.cpp:19]: = 100 cents?

File: Exercises/NoModules/Chapter 04/Soln4_08.cpp [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    // We add 0.5 to compensate for errors in binary floating-point representation
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 04/Soln4_08.cpp:20]: was this explained anywhere?

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Exercises/NoModules/Chapter 04/Soln4_08.cpp (hunk 14 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 04/Soln4_08.cpp:20]: Floating-point rounding issues are explained plenty of times in the first chapters, yes. Maybe not how I'd solve this today, but good enough I suppose...

**Pull request merged**